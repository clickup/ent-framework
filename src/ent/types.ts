import type { Schema } from "../abstract/Schema";
import type {
  CountInput,
  ExistsInput,
  ID,
  LoadByInput,
  Order,
  Table,
  UniqueKey,
  Where,
} from "../types";
import type { Validation } from "./Validation";
import type { VC } from "./VC";

/**
 * A very shallow interface of Ent class (as a collection of static methods).
 * Used in some places where we need the very minimum from the Ent.
 */
export interface EntClass<TTable extends Table = any> {
  readonly SCHEMA: Schema<TTable>;
  readonly VALIDATION: Validation<TTable>;
  readonly name: string; // class constructor name

  new (): Ent;
  loadX(vc: VC, id: string): Promise<Ent>;
  loadNullable(vc: VC, id: string): Promise<Ent | null>;
  loadIfReadableNullable(vc: VC, id: string): Promise<Ent | null>;
  count(vc: VC, where: CountInput<TTable>): Promise<number>;
  exists(vc: VC, where: ExistsInput<TTable>): Promise<boolean>;
  select(
    vc: VC,
    where: Where<TTable>,
    limit: number,
    order?: Order<TTable>
  ): Promise<Ent[]>;
  selectChunked(
    vc: VC,
    where: Where<TTable>,
    chunkSize: number,
    limit: number,
    custom?: {}
  ): AsyncIterableIterator<Ent[]>;
  loadByX(vc: VC, keys: LoadByInput<TTable, UniqueKey<TTable>>): Promise<Ent>;
}

/**
 * A very shallow interface of one Ent.
 */
export interface Ent {
  readonly [ID]: string;
  readonly vc: VC;
  deleteOriginal(): Promise<boolean>;
}
