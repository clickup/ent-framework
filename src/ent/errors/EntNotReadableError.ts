import type { RowWithID } from "../../types";
import { ID } from "../../types";
import { EntAccessError } from "./EntAccessError";

/**
 * Error: thrown when an Ent cannot be read due to privacy reasons.
 */
export class EntNotReadableError extends EntAccessError {
  constructor(
    entName: string,
    public readonly vc: string,
    public readonly row: RowWithID,
    cause: string | { message: string } | null = null
  ) {
    super(entName, `${entName}:${row[ID]} is not readable in ${vc}`, cause);
  }
}
