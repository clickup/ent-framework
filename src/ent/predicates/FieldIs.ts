import type { VC } from "../VC";
import type {
  AbstractIs,
  ValidatorPlainResult,
  ValidatorStandardSchemaResult,
  ValidatorZodSafeParseResult,
} from "./AbstractIs";
import { maybeThrowEntValidationError } from "./internal/maybeThrowEntValidationError";

/**
 * A field validator function that returns a boolean.
 */
export type FieldIsValidatorPlain<
  TField extends string,
  TRow extends Partial<Record<TField, unknown>>,
> = (
  fieldValue: TRow[TField],
  row: TRow,
  vc: VC,
) => ValidatorPlainResult | Promise<ValidatorPlainResult>;

/**
 * A field validator function that returns a Zod result.
 */
export type FieldIsValidatorZodSafeParse<TRow> = (
  fieldValue: unknown,
  row: TRow,
  vc: VC,
) => ValidatorZodSafeParseResult | Promise<ValidatorZodSafeParseResult>;

/**
 * A field validator function that returns a Standard Schema validation result.
 */
export type FieldIsValidatorStandardSchemaV1<TRow> = (
  fieldValue: unknown,
  row: TRow,
  vc: VC,
) => ValidatorStandardSchemaResult | Promise<ValidatorStandardSchemaResult>;

/**
 * Checks that the validator function returns true for the value in some field.
 */
export class FieldIs<
  TField extends string,
  TRow extends Partial<Record<TField, unknown>>,
> implements AbstractIs<TRow>
{
  readonly name;
  readonly field: TField;
  readonly message: string | null;
  readonly validator;

  /**
   * Manual validator. Implies that we can trust the fieldValue TS type.
   */
  constructor(
    field: TField,
    validator: FieldIsValidatorPlain<TField, TRow>,
    message: string,
  );

  /**
   * Rich validator, like Standard Schema (https://standardschema.dev) or Zod.
   * No implications are made on the fieldValue type.
   */
  constructor(
    field: TField,
    validator:
      | FieldIsValidatorStandardSchemaV1<TRow>
      | FieldIsValidatorZodSafeParse<TRow>,
  );

  constructor(
    field: TField,
    validator:
      | FieldIsValidatorPlain<TField, TRow>
      | FieldIsValidatorZodSafeParse<TRow>
      | FieldIsValidatorStandardSchemaV1<TRow>,
    message?: string,
  ) {
    this.name = this.constructor.name + "(" + field + ")";
    this.field = field;
    this.message = message ?? null;
    this.validator = validator;
  }

  /**
   * Returns true if validation succeeds. Returns false if it wants the client
   * to use this.message as a validation failure response. Throws an instance of
   * EntValidationError when it needs to deliver the detailed error messages
   * about multiple fields.
   */
  async check(vc: VC, row: TRow): Promise<boolean> {
    const res = await this.validator(row[this.field], row, vc);
    return maybeThrowEntValidationError({
      name: this.name,
      field: this.field,
      res,
      allowRichResult: this.message === null,
    });
  }
}
