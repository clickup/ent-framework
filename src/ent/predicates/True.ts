import { VC } from "../VC";
import { Predicate } from "./Predicate";

/**
 * Always passes; used for e.g. globally accessed objects.
 */
export class True implements Predicate<never> {
  readonly name = this.constructor.name;

  async check(_vc: VC, _row: object): Promise<boolean> {
    return true;
  }
}
