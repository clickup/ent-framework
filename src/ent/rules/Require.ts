import type { VC } from "../VC";
import type { RuleResult } from "./Rule";
import { Rule, RuleDecision } from "./Rule";

/**
 * Returns TOLERATE if the predicate succeeds, otherwise DENY.
 * - Used mostly for write permission checks.
 * - This rule may still throw an exception if it's a wild one (i.e. not derived
 *   from EntAccessError).
 */
export class Require<TInput extends object> extends Rule<TInput> {
  readonly _TAG!: "Require";

  async evaluate(vc: VC, input: TInput): Promise<RuleResult> {
    return (await this.predicate.check(vc, input))
      ? { decision: RuleDecision.TOLERATE, rule: this, cause: null }
      : { decision: RuleDecision.DENY, rule: this, cause: null };
  }
}
