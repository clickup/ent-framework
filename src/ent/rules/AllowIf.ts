import { EntAccessError } from "../errors/EntAccessError";
import type { VC } from "../VC";
import type { RuleResult } from "./Rule";
import { Rule } from "./Rule";

/**
 * Returns ALLOW if the predicate succeeds, otherwise SKIP.
 * - Used mostly for read permission checks.
 * - This rule may still throw an exception if the exception is a wild one (not
 *   derived from EntAccessError).
 */
export class AllowIf<TInput extends object> extends Rule<TInput> {
  readonly _TAG!: "AllowIf";

  async evaluate(vc: VC, input: TInput): Promise<RuleResult> {
    try {
      return (await this.predicate.check(vc, input))
        ? { decision: "ALLOW", rule: this, cause: null }
        : { decision: "SKIP", rule: this, cause: null };
    } catch (error: unknown) {
      if (error instanceof EntAccessError) {
        // We carry a cause for this SKIP decision too if it was due to an
        // access-related error.
        return { decision: "SKIP", rule: this, cause: error };
      }

      throw error;
    }
  }
}
