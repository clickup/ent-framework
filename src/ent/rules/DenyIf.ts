import { EntAccessError } from "../errors/EntAccessError";
import type { VC } from "../VC";
import type { RuleResult } from "./Rule";
import { Rule, RuleDecision } from "./Rule";

/**
 * Returns DENY if the predicate succeeds, otherwise SKIP.
 * - Used mostly to early block some read/write access.
 * - EntAccessError exception will be treated as a DENY signal (so it will abort
 *   processing immediately).
 * - This rule may still throw an exception if the exception is a wild one (not
 *   derived from EntAccessError).
 */
export class DenyIf<TInput extends object> extends Rule<TInput> {
  async evaluate(vc: VC, input: TInput): Promise<RuleResult> {
    try {
      return (await this.predicate.check(vc, input))
        ? { decision: RuleDecision.DENY, rule: this, cause: null }
        : { decision: RuleDecision.SKIP, rule: this, cause: null };
    } catch (error) {
      if (error instanceof EntAccessError) {
        // We carry a cause for this DENY decision too if it was due to an
        // access-related error.
        return { decision: RuleDecision.DENY, rule: this, cause: error };
      }

      throw error;
    }
  }
}
