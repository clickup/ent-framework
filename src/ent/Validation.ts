import { indent } from "../helpers";
import {
  InsertFieldsRequired,
  InsertInput,
  Row,
  Table,
  UpdateInput,
} from "../types";
import { EntNotInsertableError } from "./errors/EntNotInsertableError";
import { EntNotReadableError } from "./errors/EntNotReadableError";
import { EntNotUpdatableError } from "./errors/EntNotUpdatableError";
import {
  EntValidationError,
  EntValidationErrorInfo,
} from "./errors/EntValidationError";
import { Predicate } from "./predicates/Predicate";
import { Require } from "./rules/Require";
import { evaluate, Rule, RuleDecision, RuleResult } from "./rules/Rule";
import { buildNewRow } from "./Triggers";
import { VC } from "./VC";

export type ValidationRules<TTable extends Table> = {
  readonly tenantUserIDField?: InsertFieldsRequired<TTable> & string;
  readonly load: Validation<TTable>["load"];
  readonly insert: Validation<TTable>["insert"];
  readonly update?: Validation<TTable>["update"];
  readonly delete?: Validation<TTable>["delete"];
  readonly validate?: Array<Predicate<Row<TTable>> & EntValidationErrorInfo>;
};

export class Validation<TTable extends Table> {
  readonly tenantUserIDField?: InsertFieldsRequired<TTable> & string;
  readonly load: Array<Rule<Row<TTable>>>;
  readonly insert: Array<Rule<InsertInput<TTable>>>;
  readonly update: Array<Rule<Row<TTable>>>;
  readonly delete: Array<Rule<Row<TTable>>>;
  readonly validate: Array<Require<Row<TTable>>>;

  constructor(private entName: string, rules: ValidationRules<TTable>) {
    this.load = rules.load;
    this.insert = rules.insert;
    this.update = rules.update || (this.insert as any);
    this.delete = rules.delete || this.update;
    this.validate = (rules.validate || []).map((pred) => new Require(pred));
    this.tenantUserIDField = rules.tenantUserIDField;
  }

  async validateLoad(vc: VC, row: Row<TTable>): Promise<void> {
    await this.validatePrivacyImpl(
      "load",
      this.load,
      vc,
      row,
      false,
      EntNotReadableError
    );
  }

  async validateInsert(vc: VC, input: InsertInput<TTable>): Promise<void> {
    await this.validateUserInputImpl(vc, input as Row<TTable>);
    await this.validatePrivacyImpl(
      "insert",
      this.insert,
      vc,
      input,
      true, // parallel
      EntNotInsertableError
    );
  }

  async validateUpdate(
    vc: VC,
    old: Row<TTable>,
    input: UpdateInput<TTable>,
    privacyOnly = false
  ): Promise<void> {
    // Simulate the update, as if it's applied to the ent.
    const newRow = buildNewRow(old, input);

    if (!privacyOnly) {
      await this.validateUserInputImpl(vc, newRow);
    }

    await this.validatePrivacyImpl(
      "update",
      this.update,
      vc,
      newRow,
      true, // parallel
      EntNotUpdatableError
    );
  }

  async validateDelete(vc: VC, row: Row<TTable>): Promise<void> {
    await this.validatePrivacyImpl(
      "delete",
      this.delete,
      vc,
      row,
      true, // parallel
      EntNotUpdatableError // same exception as for update
    );
  }

  private async validatePrivacyImpl(
    op: string,
    rules: Array<Rule<object>>,
    vc: VC,
    row: object,
    parallel: boolean,
    ExceptionClass:
      | typeof EntNotReadableError
      | typeof EntNotInsertableError
      | typeof EntNotUpdatableError
  ) {
    this.validateTenantUserIDImpl(vc, row, ExceptionClass);

    const { allow, results } = await evaluate(rules, vc, row, parallel);
    if (allow) {
      return;
    }

    throw new ExceptionClass(this.entName, vc.toString(), row as any, {
      message: resultsToText(op, results),
    });
  }

  private validateTenantUserIDImpl(
    vc: VC,
    row: object,
    ExceptionClass:
      | typeof EntNotReadableError
      | typeof EntNotInsertableError
      | typeof EntNotUpdatableError
  ) {
    if (this.tenantUserIDField === undefined) {
      return;
    }

    const rowTenantUserID = (row as any)[this.tenantUserIDField];
    if (rowTenantUserID === vc.userID) {
      return;
    }

    throw new ExceptionClass(this.entName, vc.toString(), row as any, {
      message:
        this.tenantUserIDField +
        " is expected to be " +
        JSON.stringify(vc.userID) +
        ", but got " +
        JSON.stringify(rowTenantUserID),
    });
  }

  private async validateUserInputImpl(vc: VC, row: Row<TTable>) {
    const { allow, results } = await evaluate(this.validate, vc, row, true);
    if (allow) {
      return;
    }

    const failedPredicates = results
      .filter((result) => result.decision === RuleDecision.DENY)
      .map((result) => result.rule.predicate as any as EntValidationErrorInfo);
    if (failedPredicates.length > 0) {
      throw new EntValidationError(this.entName, failedPredicates);
    }
  }
}

/**
 * A helper function which returns a debugging text for a list of rule
 * evaluation results.
 */
function resultsToText(op: string, results: RuleResult[]) {
  if (results.length === 0) {
    return `No "${op}" rules defined`;
  }

  return results
    .map(
      ({ rule, decision, cause }) =>
        "Rule " +
        rule.name +
        " returned " +
        decision +
        (cause ? ", because:\n" + indent(cause.message) : "")
    )
    .join("\n");
}
