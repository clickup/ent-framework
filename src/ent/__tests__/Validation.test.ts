import { ID } from "../../types";
import { EntNotReadableError } from "../errors/EntNotReadableError";
import { FieldIs } from "../predicates/FieldIs";
import { True } from "../predicates/True";
import { AllowIf } from "../rules/AllowIf";
import { DenyIf } from "../rules/DenyIf";
import { Require } from "../rules/Require";
import { Validation } from "../Validation";
import { createVC } from "./helpers/test-objects";
import ValidationTester from "./helpers/ValidationTester";

const companyTable = {
  id: { type: ID },
  tenant_id: { type: ID },
} as const;

const vc = createVC();

test("0000: load succeeds when first rule allows", async () => {
  const tester = new ValidationTester();
  await tester.matchSnapshot(
    new Validation<typeof companyTable>("table", {
      load: [
        new AllowIf(async function First(_vc, _row) {
          return tester.respond("First", true);
        }),
        new AllowIf(async function Second(_vc, _row) {
          return tester.respond("Second", false);
        }),
      ],
      insert: [],
    }),
    { id: "123", tenant_id: "42" },
    "validateLoad",
    undefined,
    vc.toLowerInternal("42")
  );
});

test("0010: load succeeds when any rule allows", async () => {
  const tester = new ValidationTester();
  await tester.matchSnapshot(
    new Validation<typeof companyTable>("table", {
      load: [
        new AllowIf(async function First(_vc, _row) {
          return tester.respond("First", false);
        }),
        new AllowIf(async function Second(_vc, _row) {
          return tester.respond("Second", true);
        }),
      ],
      insert: [],
    }),
    { id: "123", tenant_id: "42" }
  );
});

test("0020: load fails when first rule throws", async () => {
  const tester = new ValidationTester();
  await tester.matchSnapshot(
    new Validation<typeof companyTable>("table", {
      load: [
        new AllowIf(async function First(_vc, _row) {
          return tester.respond("First", Error("wild"));
        }),
      ],
      insert: [],
    }),
    { id: "123", tenant_id: "42" }
  );
});

test("0030: insert succeeds when all require allow", async () => {
  const tester = new ValidationTester();
  await tester.matchSnapshot(
    new Validation<typeof companyTable>("table", {
      load: [],
      insert: [
        new Require(async function First(_vc, row) {
          return tester.respond("First", true, row);
        }),
        new Require(async function Second(_vc, row) {
          return tester.respond("Second", true, row);
        }),
      ],
    }),
    { id: "123", tenant_id: "42" },
    "validateInsert",
    undefined,
    vc.toLowerInternal("42")
  );
});

test("0040: insert fails when any require denies", async () => {
  const tester = new ValidationTester();
  await tester.matchSnapshot(
    new Validation<typeof companyTable>("table", {
      load: [],
      insert: [
        new Require(async function First(_vc, _row) {
          return tester.respond("First", true);
        }),
        new Require(async function Second(_vc, _row) {
          return tester.respond("Second", false);
        }),
      ],
    }),
    { id: "123", tenant_id: "42" }
  );
});

test("0041: update fails when any require denies", async () => {
  const tester = new ValidationTester();
  await tester.matchSnapshot(
    new Validation<typeof companyTable>("table", {
      load: [],
      insert: [],
      update: [
        new Require(async function First(_vc, _row) {
          return tester.respond("First", true);
        }),
        new Require(async function Second(_vc, _row) {
          return tester.respond("Second", false);
        }),
      ],
    }),
    { id: "123", tenant_id: "42" },
    "validateUpdate",
    undefined,
    vc.toLowerInternal("42")
  );
});

test("0042: delete fails when any require denies", async () => {
  const tester = new ValidationTester();
  await tester.matchSnapshot(
    new Validation<typeof companyTable>("table", {
      load: [],
      insert: [],
      delete: [
        new Require(async function First(_vc, _row) {
          return tester.respond("First", true);
        }),
        new Require(async function Second(_vc, _row) {
          return tester.respond("Second", false);
        }),
      ],
    }),
    { id: "123", tenant_id: "42" },
    "validateDelete",
    undefined,
    vc.toLowerInternal("42")
  );
});

test("0043: update fails when user errors", async () => {
  const tester = new ValidationTester();
  await tester.matchSnapshot(
    new Validation<typeof companyTable>("table", {
      load: [],
      insert: [new AllowIf(new True())],
      validate: [
        new FieldIs("tenant_id", (_value) => false, "some one"),
        new FieldIs("tenant_id", (_value) => false, "some two"),
      ],
    }),
    { id: "123", tenant_id: "42" },
    "validateUpdate",
    { tenant_id: "42" }
  );
});

test("0044: update succeeds user validation when field untouched", async () => {
  const tester = new ValidationTester();
  await tester.matchSnapshot(
    new Validation<typeof companyTable>("table", {
      load: [],
      insert: [new AllowIf(new True())],
      validate: [new FieldIs("id", (_value) => false, "some one")],
    }),
    { id: "123", tenant_id: "42" },
    "validateUpdate",
    { tenant_id: "101" }
  );
});

test("0050: insert fails when any require throws", async () => {
  const tester = new ValidationTester();
  await tester.matchSnapshot(
    new Validation<typeof companyTable>("table", {
      load: [],
      insert: [
        new Require(async function First(_vc, _row) {
          return tester.respond("First", true);
        }),
        new Require(async function Second(_vc, _row) {
          return tester.respond("Second", Error("wild"));
        }),
      ],
    }),
    { id: "123", tenant_id: "42" }
  );
});

test("0060: load succeeds when any rule allows even if another rule throws EntNotReadableError", async () => {
  const tester = new ValidationTester();
  await tester.matchSnapshot(
    new Validation<typeof companyTable>("table", {
      load: [
        new AllowIf(async function First(vc, _row) {
          return tester.respond(
            "First",
            // Not a "wild" exception (since derived from EntAccessError).
            new EntNotReadableError(
              "other_table",
              vc.toString(),
              { id: "987" },
              { message: "ent access error" }
            )
          );
        }),
        new AllowIf(async function Second(_vc, _row) {
          return tester.respond("Second", true);
        }),
      ],
      insert: [],
    }),
    { id: "123", tenant_id: "42" }
  );
});

test("0070: load fails when any rule allows but another rule throws any wild exception", async () => {
  const tester = new ValidationTester();
  await tester.matchSnapshot(
    new Validation<typeof companyTable>("table", {
      load: [
        new AllowIf(async function First(_vc, _row) {
          // "Wild" means "not derived from EntAccessError"
          return tester.respond("First", Error("wild"));
        }),
        new AllowIf(async function Second(_vc, _row) {
          return tester.respond("Second", true);
        }),
      ],
      insert: [],
    }),
    { id: "123", tenant_id: "42" }
  );
});

test("0080: validations fail when no rules defined", async () => {
  const tester = new ValidationTester();
  await tester.matchSnapshot(
    new Validation<typeof companyTable>("table", {
      load: [],
      insert: [],
    }),
    { id: "123", tenant_id: "42" },
    "validateLoad"
  );
  await tester.matchSnapshot(
    new Validation<typeof companyTable>("table", {
      load: [],
      insert: [],
    }),
    { id: "123", tenant_id: "42" },
    "validateInsert"
  );
});

test("0090: load fails with nice error message if only one rule", async () => {
  const tester = new ValidationTester();
  await tester.matchSnapshot(
    new Validation<typeof companyTable>("table", {
      load: [
        new AllowIf(async function First(_vc, _row) {
          return tester.respond("First", Error("wild"));
        }),
      ],
      insert: [],
    }),
    { id: "123", tenant_id: "42" }
  );
  await tester.matchSnapshot(
    new Validation<typeof companyTable>("table", {
      load: [
        new AllowIf(async function First(_vc, _row) {
          return tester.respond("First", false);
        }),
      ],
      insert: [],
    }),
    { id: "123", tenant_id: "42" }
  );
});

test("0100: insert fails with nice error message if only one rule", async () => {
  const tester = new ValidationTester();
  await tester.matchSnapshot(
    new Validation<typeof companyTable>("table", {
      load: [],
      insert: [
        new AllowIf(async function First(_vc, _row) {
          return tester.respond("First", Error("wild"));
        }),
      ],
    }),
    { id: "123", tenant_id: "42" }
  );
  await tester.matchSnapshot(
    new Validation<typeof companyTable>("table", {
      load: [],
      insert: [
        new AllowIf(async function First(_vc, _row) {
          return tester.respond("First", false);
        }),
      ],
    }),
    { id: "123", tenant_id: "42" }
  );
});

test("0110: insert succeeds when DenyIf rule evaluates", async () => {
  const tester = new ValidationTester();
  await tester.matchSnapshot(
    new Validation<typeof companyTable>("table", {
      load: [],
      insert: [
        new DenyIf(async function First(_vc, _row) {
          return tester.respond("First", false);
        }),
        new AllowIf(new True()),
      ],
    }),
    { id: "123", tenant_id: "42" }
  );
});

test("0120: load fails when DenyIf rule throws", async () => {
  const tester = new ValidationTester();
  await tester.matchSnapshot(
    new Validation<typeof companyTable>("table", {
      load: [
        new DenyIf(async function First(vc, _row) {
          return tester.respond(
            "First",
            new EntNotReadableError(
              "other_table",
              vc.toString(),
              { id: "987" },
              { message: "ent access error" }
            )
          );
        }),
      ],
      insert: [],
    }),
    { id: "123", tenant_id: "42" }
  );
});

test("0130: fail when tenant user id mismatches", async () => {
  const tester = new ValidationTester();
  const validation = new Validation<typeof companyTable>("table", {
    tenantPrincipalField: "tenant_id",
    load: [],
    insert: [],
  });
  await tester.matchSnapshot(
    validation,
    { id: "123", tenant_id: "42" },
    "validateLoad",
    undefined,
    vc.toLowerInternal("999")
  );
  await tester.matchSnapshot(
    validation,
    { id: "123", tenant_id: "42" },
    "validateInsert",
    undefined,
    vc.toLowerInternal("999")
  );
  await tester.matchSnapshot(
    validation,
    { id: "123" } as any,
    "validateInsert",
    undefined,
    vc.toLowerInternal("999")
  );
  await tester.matchSnapshot(
    validation,
    { id: "123", tenant_id: "42" },
    "validateUpdate",
    undefined,
    vc.toLowerInternal("999")
  );
  await tester.matchSnapshot(
    validation,
    { id: "123", tenant_id: "42" },
    "validateDelete",
    undefined,
    vc.toLowerInternal("999")
  );
});
