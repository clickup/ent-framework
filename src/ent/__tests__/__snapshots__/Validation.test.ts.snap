// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`0000: load succeeds when first rule allows 1`] = `
"OK
--- (what actually happened) ---
[First returned] true"
`;

exports[`0010: load succeeds when any rule allows 1`] = `
"OK
--- (what actually happened) ---
[First returned] false
[Second returned] true"
`;

exports[`0020: load fails when first rule throws 1`] = `
"Failure
--- (error returned to client) ---
Error: wild
--- (what actually happened) ---
[First threw] Error: wild"
`;

exports[`0030: insert succeeds when all require allow 1`] = `
"OK
--- (what actually happened) ---
[First { id: '123', tenant_id: '42' } returned] true
[Second { id: '123', tenant_id: '42' } returned] true"
`;

exports[`0040: insert fails when any require denies 1`] = `
"Failure
--- (error returned to client) ---
EntNotInsertableError: table: cannot insert in vc:guest:master, because:
  Rule Require:First returned TOLERATE
  Rule Require:Second returned DENY
--- (what actually happened) ---
[First returned] true
[Second returned] false"
`;

exports[`0041: update fails when any require denies 1`] = `
"Failure
--- (error returned to client) ---
EntNotUpdatableError: table:123 is not updatable/deletable in vc:42, because:
  Rule Require:First returned TOLERATE
  Rule Require:Second returned DENY
--- (what actually happened) ---
[First returned] true
[Second returned] false"
`;

exports[`0042: delete fails when any require denies 1`] = `
"Failure
--- (error returned to client) ---
EntNotUpdatableError: table:123 is not updatable/deletable in vc:42, because:
  Rule Require:First returned TOLERATE
  Rule Require:Second returned DENY
--- (what actually happened) ---
[First returned] true
[Second returned] false"
`;

exports[`0043: update fails when user errors 1`] = `
"Failure
--- (error returned to client) ---
EntValidationError: table: tenant_id: "some one", tenant_id: "some two"
--- (what actually happened) ---
"
`;

exports[`0044: update succeeds user validation when field untouched 1`] = `
"OK
--- (what actually happened) ---
"
`;

exports[`0050: insert fails when any require throws 1`] = `
"Failure
--- (error returned to client) ---
Error: wild
--- (what actually happened) ---
[First returned] true
[Second threw] Error: wild"
`;

exports[`0060: load succeeds when any rule allows even if another rule throws EntNotReadableError 1`] = `
"OK
--- (what actually happened) ---
[First threw] EntNotReadableError: other_table:987 is not readable in vc:guest:master, because:
  ent access error
[Second returned] true"
`;

exports[`0070: load fails when any rule allows but another rule throws any wild exception 1`] = `
"Failure
--- (error returned to client) ---
Error: wild
--- (what actually happened) ---
[First threw] Error: wild"
`;

exports[`0080: validations fail when no rules defined 1`] = `
"Failure
--- (error returned to client) ---
EntNotReadableError: table:123 is not readable in vc:guest:master, because:
  No "load" rules defined
--- (what actually happened) ---
"
`;

exports[`0080: validations fail when no rules defined 2`] = `
"Failure
--- (error returned to client) ---
EntNotInsertableError: table: cannot insert in vc:guest:master, because:
  No "insert" rules defined
--- (what actually happened) ---
"
`;

exports[`0090: load fails with nice error message if only one rule 1`] = `
"Failure
--- (error returned to client) ---
Error: wild
--- (what actually happened) ---
[First threw] Error: wild"
`;

exports[`0090: load fails with nice error message if only one rule 2`] = `
"Failure
--- (error returned to client) ---
EntNotReadableError: table:123 is not readable in vc:guest:master, because:
  Rule AllowIf:First returned SKIP
--- (what actually happened) ---
[First threw] Error: wild
[First returned] false"
`;

exports[`0100: insert fails with nice error message if only one rule 1`] = `
"Failure
--- (error returned to client) ---
Error: wild
--- (what actually happened) ---
[First threw] Error: wild"
`;

exports[`0100: insert fails with nice error message if only one rule 2`] = `
"Failure
--- (error returned to client) ---
EntNotInsertableError: table: cannot insert in vc:guest:master, because:
  Rule Require:First returned DENY
--- (what actually happened) ---
[First threw] Error: wild
[First returned] false"
`;

exports[`0110: insert succeeds when DenyIf rule evaluates 1`] = `
"OK
--- (what actually happened) ---
[First returned] false"
`;

exports[`0120: load fails when DenyIf rule throws 1`] = `
"Failure
--- (error returned to client) ---
EntNotReadableError: table:123 is not readable in vc:guest:master, because:
  Rule DenyIf:First returned DENY, because:
    other_table:987 is not readable in vc:guest:master, because:
      ent access error
--- (what actually happened) ---
[First threw] EntNotReadableError: other_table:987 is not readable in vc:guest:master, because:
  ent access error"
`;

exports[`0130: fail when tenant user id mismatches 1`] = `
"Failure
--- (error returned to client) ---
EntNotReadableError: table:123 is not readable in vc:999, because:
  tenant_id is expected to be "999", but got "42"
--- (what actually happened) ---
"
`;

exports[`0130: fail when tenant user id mismatches 2`] = `
"Failure
--- (error returned to client) ---
EntNotInsertableError: table: cannot insert in vc:999, because:
  tenant_id is expected to be "999", but got "42"
--- (what actually happened) ---
"
`;

exports[`0130: fail when tenant user id mismatches 3`] = `
"Failure
--- (error returned to client) ---
EntNotInsertableError: table: cannot insert in vc:999, because:
  tenant_id is expected to be "999", but got undefined
--- (what actually happened) ---
"
`;

exports[`0130: fail when tenant user id mismatches 4`] = `
"Failure
--- (error returned to client) ---
EntNotUpdatableError: table:123 is not updatable/deletable in vc:999, because:
  tenant_id is expected to be "999", but got "42"
--- (what actually happened) ---
"
`;

exports[`0130: fail when tenant user id mismatches 5`] = `
"Failure
--- (error returned to client) ---
EntNotUpdatableError: table:123 is not updatable/deletable in vc:999, because:
  tenant_id is expected to be "999", but got "42"
--- (what actually happened) ---
"
`;

exports[`0140: load succeeds when some of Or predicates succeed 1`] = `
"OK
--- (what actually happened) ---
[First threw] EntNotReadableError: other_table:987 is not readable in vc:guest, because:
  ent access error
[Second returned] true"
`;

exports[`0150: load fails with nice error when all of Or predicates fail 1`] = `
"Failure
--- (error returned to client) ---
EntNotInsertableError: table: cannot insert in vc:guest:master, because:
  Rule Require:Or returned DENY, because:
    First: other_table:987 is not readable in vc:guest, because:
      { myKey: 'ent access error' }
    Second: false
--- (what actually happened) ---
[First threw] EntNotReadableError: other_table:987 is not readable in vc:guest, because:
  { myKey: 'ent access error' }
[Second returned] false"
`;

exports[`0160: load crashes when some predicates fail with a wild error 1`] = `
"Failure
--- (error returned to client) ---
Error: wild
--- (what actually happened) ---
[First threw] Error: wild
[Second returned] true"
`;
